test/forms/WhatIsGIINFormProviderSpec.scala

```diff
@@ -21,34 +21,67 @@ import play.api.data.FormError
 
 class WhatIsGIINFormProviderSpec extends StringFieldBehaviours {
 
-  val requiredKey = "whatIsGIIN.error.required"
-  val lengthKey   = "whatIsGIIN.error.length"
-  val maxLength   = 100
-
   val form = new WhatIsGIINFormProvider()()
 
   ".value" - {
 
-    val fieldName = "value"
+    val fieldName      = "value"
+    val requiredKey    = "whatIsGIIN.error.required"
+    val lengthKey      = "whatIsGIIN.error.length"
+    val invalidKey     = "whatIsGIIN.error.invalid"
+    val formatKey      = "whatIsGIIN.error.format"
+    val invalidCharKey = "whatIsGIIN.error.char"
+
+    val giinSetLength = 19
 
     behave like fieldThatBindsValidData(
       form,
       fieldName,
-      stringsWithMaxLength(maxLength)
+      validGIIN
     )
 
-    behave like fieldWithMaxLength(
+    behave like fieldWithMaxLengthAlpha(
       form,
       fieldName,
-      maxLength = maxLength,
-      lengthError = FormError(fieldName, lengthKey, Seq(maxLength))
+      maxLength = giinSetLength,
```

Comment:
Trying to get my head around this... so this test is generating a string with a maximum length of 19 characters? Does that mean that one in 19 times it could generate a 19 character string which would actually be the correct length for a GIIN?

Does the maxLength then need to be `giinSetLength - 1` so that it will always be smaller than 19 characters?

Reply:
No it takes the now "setLength" value of 19, and inputs that value into the stringsLongerThanAlpha generator method which generates a random number between 20 (or setLength + 1) and 100. It then compares whether that value is larger than the setLength and checks if the lengthKey error is returned if it is.

Reply:
Sorry, need more coffee. I get it now. The behaviour is `fieldWithMaxLength` so it makes sense it would use generators to create values longer than this max length. Then the inverse for the min length. Get it now.